src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs

```diff
@@ -14180,5 +14180,102 @@ .maxstack  1
             verifier.VerifyIL("NativeInts.Checked4", expectedCheckedIL);
             verifier.VerifyIL("NativeInts.Checked5", expectedCheckedIL);
         }
+
+        [Fact, WorkItem(53306, "https://github.com/dotnet/roslyn/issues/53306")]
+        public void TestNativeIntegerOperatorsInIntPtrAndUIntPtr()
+        {
+            var source =
+@"namespace System
+{
+    public class Object { }
+    public class String { }
+    public struct Boolean { }
+    public class Enum : ValueType { }
+    public struct Void { }
+    public class Attribute { }
+    public abstract class ValueType { }
+
+    public sealed class AttributeUsageAttribute : Attribute
+    { 
+        public AttributeUsageAttribute(AttributeTargets validOn)
+        {
+        }
+
+        public bool AllowMultiple { get; set; }
+        public bool Inherited { get; set; }
+    }
+
+    public enum AttributeTargets { }
+
+    public struct Int32 { }
+    public struct IntPtr
+    {
+        public static nint operator+(nint a, nint b) => (nint)(a + b);
```

Comment:
>operator+(nint a, nint b) => (nint)(a + b) [](http://example.com/codeflow?start=27&length=42)

We should test whether an expression `a + b` binds to this `operator+(nint, nint)`. I suspect we won't bind to it (by design) because addition of `nint` arguments is handled directly in the compiler, but we should confirm that.

Please also consider testing binding to an explicit operator with arguments that are not directly handled by the compiler - say `operator+(nint a, bool b)`.

Reply:
`IntPtr + IntPtr` is also interesting to consider.

Once we start implementing the generic math interfaces, `IntPtr + IntPtr` will become "legal" and if `unchecked` operators are approved, we will have `checked(IntPtr + IntPtr)` be valid as well, at which point `nint` and `IntPtr` will be able to be identical (assuming the framework takes the breaking change for the existing `IntPtr + int` operators).

Reply:
@cston The IL for the operator is:

```il
    {
      // Code size        4 (0x4)
      .maxstack  2
      IL_0000:  ldarg.0
      IL_0001:  ldarg.1
      IL_0002:  add
      IL_0003:  ret
    }
```

That seems correct that it doesn't bind to the defined operator, right?

Reply:
> IntPtr + IntPtr is also interesting to consider.

Did that mean to test the binding of an `IntPtr + IntPtr`? e.g:

```csharp
public void M(IntPtr a, IntPtr b) => _ = a + b; // what the + will bind to?
```

Or it meant to define an operator for addition of two `IntPtr`s? (If that's not allowed today - Then I'm not sure why it's not, or why my change would affect (or would need to affect) that.

(Sorry I don't have enough experience around using those types)

Reply:
We should have tests that cover explicitly defined unary and binary operators with `nint` arguments, and test use of those operators with combinations of { `IntPtr`, `IntPtr?`, `nint`, `nint?` }.

Specifically, we should test:
- operators that are directly supported in the compiler: `operator-(nint i)`, `operator*(nint x, nint y)`
- operators that are not directly supported in the compiler: `operator!(nint i)`, `operator*(nint x, bool y)`, `operator*(bool x, nint y)`

We might need to test `implicit` and `explicit` conversion operators as well.